### DISCUSSION TITLE
[Chapter 06] 부정의 의미가 포함된 메서드 네이밍

### DISCUSSION THREAD

**Discussion by corock** [cowaine/is-my-code-that-strange#44] [Chapter 06 (조건 분기)] [answered]
89 페이지에는 조기 반환(_early return_)을 소개하면서 `Member` 클래스의 `canAct()` 메서드를 아래와 같이 수정했습니다.

```java
if (!member.canAct()) {
    // (...)
}
```

하지만 어떤 개발자는 다음처럼 리팩터링할 수도 있지 않을까요?

```java
if (member.cannotAct()) {
    // (...)
}
```

NOT 논리 연산자인 `!` 연산자를 사용하는 것보다 메서드명 자체로 그 의미를 나타내도록 네이밍한 상황을 예시로 들어봤습니다.

만약 여러분이 어떤 팀을 이끄는 리더일 때, 팀원 A, B 가 위에 소개한 두 방식을 주장하면서 첨예하게 대립한다면 어떤 결정을 내리실 건가요?
이 결정은 여러분의 선호하는 방식으로 결정된다고 치고, 왜 그 방식을 선호하는지 간단하게 댓글로 남겨주시면 좋을 것 같습니다.

**Comment by snoopingh** [comment]
긍정적 의미가 포함된 메서드를 더 선호합니다.
부정적 의미로 할 경우 조건문에서 부정을 또 해야 하는 경우 (이중 부정) 가독성이 떨어지는 것 같습니다.

**Comment by Irisation23** [comment, marked as answer]
### 메서드 네이밍이 부정을 담고 있는 것 보다는 긍정문을 통해 `!` 를 이용하여 논리를 전개하는게 더 좋아보입니다.

총 두가지 이유가 있습니다.

1. 자바에서 사용되는 리턴 타입이 부울 타입인 메서드의 네이밍 대부분은 긍정문으로 대부분 구성되어 있습니다. (isEmpty, isBlank)
2. 부정문에 `!`가 들어간다면 인간의 뇌가 인지하는데는 불편함을 초래합니다. (@snoopingh 과 의견이 같습니다.)
  - https://www.ilyosisa.co.kr/news/article.html?no=235184

``` java
if (!member.cannotAct()) {
    // (...)
}
```

- 위의 if을 구문대로 해석해보면 `회원이 움직일 수 없는가?` 에 대한 부정문인 `회원이 움직일 수 있는가?` 입니다.
- cannotAct()가 반환하는 true, false에 따라 해석할 수 있는 여지가 다르지만 `cannotAct()`의 네이밍이 부정문이라는 점에서 인식시점에 `!`를 달고 생각을 해야 한다는 점이 아쉬운 점으로 지목되는 이유인 것 같습니다.

결과적으로 ! 를 한번 더 생각해야하는 부정문 같은 경우는 사용하기 전에 한번 더 생각해 보는 것도 좋을 것 같습니다.

> 여담으로 첨예한 대립을 이어나가고 있다면 지켜보는게 즐거울 것 같습니다. 😁

**Comment by firewoody237** [comment]
제가 리더라면.. 쓰고싶은대로 쓰라고 할 것 같습니다.

일단 `!`가 들어감으로써 머릿속에 생각의 논리 하나가 추가가 되는건 안좋다고 생각합니다.
`member.isGoldMember`와 `member.isNotGoldMember` 이런건 좀 직관적이라 판단하기 쉽다고 생각되는데,
정규식을 사용하거나 할 때 `Pattern.match("Pattern")`과 같은경우라던가,
이미 다른 곳에서 `.isXX()` 메서드를 사용하는 곳들이 많은데, 이 메서드의 부정형을 사용하고자 하는 경우, 굳이 Wrapping하지 않고 그냥 쓰는 것 같습니다.
(추후 많아질 때 리팩토링)
